libs/authentication/api/src/lib/optional-authentication.guard.ts

```diff
@@ -0,0 +1,27 @@
+import {AuthGuard} from '@nestjs/passport'
+import {ExecutionContext, UnauthorizedException} from '@nestjs/common'
+import {GqlExecutionContext} from '@nestjs/graphql'
+
+/**
+ * This Guard is to attach the user to the request
+ */
+export class OptionalAuthenticationGuard extends AuthGuard('session') {
+  public getRequest(context: ExecutionContext) {
+    const ctx = GqlExecutionContext.create(context)
+    return ctx.getContext().req
+  }
+
+  public async canActivate(context: ExecutionContext): Promise<boolean> {
+    try {
+      await super.canActivate(context)
+    } catch (e) {
+      if (e instanceof UnauthorizedException) {
+        return true
```

Comment:
return true if user is not logged in as we don't mind. Other errors should still throw

Reply:
Why there is `UnauthorizedException` when we just try to authenticate.
Plus authentication is actually optional.

Reply:
Where is this exception thrown?

Reply:
Just checked how I did that in one project. Actually it's not so great. I check "public override"

```
canActivate(context: ExecutionContext) {
    const isPublic = this.reflector.getAllAndOverride<boolean>(IS_PUBLIC_KEY, [
      context.getHandler(),
      context.getClass(),
    ]);
    if (isPublic) {
      return true;
    }
    return super.canActivate(context);
  }
```

Reply:
That exception is thrown in the nestjs internals because the passport strategy http://www.passportjs.org/packages/passport-http-bearer/ returned false on the validate function.

This guard is only here to attach the user object to the request and nothing more.

Reply:
For me now it seems we might do it wrong.

Guard is there to protect something from access (kinda). While we are actually extracting user from session if it exists.

Theoretically we should have:
1. Pipe (or such) which transform request enriching it with user if it exists in session.
2. Explicitly specified Guard to prevent request without specific permission (based on optional user within request).

Reply:
That's as far as I have read up the recommended way by Nestjs and best practice  in express to attach the user to the request.